src/backend_rest.c

```
@@ -0,0 +1,222 @@
+/* backend_rest.c
+   Rémi Attab (remi.attab@gmail.com), 09 Jun 2016
+   FreeBSD-style copyright and disclaimer apply
+*/
+
+#include "optics.h"
+#include "utils/errors.h"
+#include "utils/lock.h"
+#include "utils/buffer.h"
+#include "utils/crest/crest.h"
+
+#include <string.h>
+
+
+// -----------------------------------------------------------------------------
+// metrics
+// -----------------------------------------------------------------------------
+
+struct metric
+{
+    char *key;
+    enum optics_lens_type type;
+    union optics_poll_value value;
+};
+
+struct metrics
+{
+    size_t len;
+    size_t cap;
+    struct metric data[];
+};
+
+enum { metrics_init_cap = 128 };
+
+static void metrics_free(struct metrics *metrics)
+{
+    if (!metrics) return;
+
+    for (size_t i = 0; i < metrics->len; ++i)
+        free(metrics->data[i].key);
+
+    free(metrics);
+}
+
+static struct metrics *metrics_append(struct metrics *metrics, const struct optics_poll *poll)
+{
+    if (!metrics) {
+        metrics = malloc(sizeof(struct metrics) + sizeof(struct metric) * metrics_init_cap);
```

The policy in iplists (with the exception of some testing code) is assert on NULL return from malloc, in case we disable overcommit.  (And use reallocf rather than realloc, asserting against NULL on its return.)

The general philosophy is any piece should fail fast in situations too complicated for it to deal with, because we know we have supervision at a higher level and the service is best-effort.

This is why I don't write libraries: what right does a library have to decide the failure behavior of the application that uses it?  To be pragmatic, though, the best you can do is clearly document what your choice was so your users don't get a nasty surprise.

Especially since overcommit is generally enabled, I feel like sqlite-style Herculean efforts to deal with allocator failure are a waste of programmer time, on the systems we work on.

Assert-on-NULL is a little problematic, too; since assertions can be accidentally disabled, it should be invoke-error-handler-on-NULL that defaults to abort() with as much info as possible.